src/app/clusters/service-area-server/service-area-delegate.h

```diff
@@ -0,0 +1,424 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "service-area-cluster-objects.h"
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ServiceArea {
+
+class Instance;
+
+// *****************************************************************************
+// cluster constraints
+
+constexpr size_t kMaxNumSupportedLocations = 255;
+constexpr size_t kMaxNumSupportedMaps      = 255;
+constexpr size_t kMaxNumSelectedLocations  = 255;
+constexpr size_t kMaxNumProgressElements   = 255;
+
+constexpr size_t kMaxSizeStatusText = 256;
+
+/**
+ * Uncopyable is a helper to ensure that Service Area cluster instances and delegates
+ * are unique and cannot be copied.
+ */
+class Uncopyable
+{
+protected:
+    Uncopyable() {}
+    ~Uncopyable() = default;
+
+private:
+    Uncopyable(const Uncopyable &)             = delete;
+    Uncopyable & operator=(const Uncopyable &) = delete;
+};
+
+/**
+ * ServiceArea::Delegate Defines methods for implementing application-specific
+ * logic for the Service Area Cluster.
+ */
+class Delegate : public Uncopyable
+{
+public:
+    Delegate()          = default;
+    virtual ~Delegate() = default;
+
+    friend class Instance;
+
+protected:
```

Comment:
why are all members in the delegate protected? This seems to make unit testing harder.

Reply:
We want to stop the use of delegate methods to modify attributes as these will not have the necessary checks.

Reply:
No. There is no unit testing. Neither in any of the other cultures that I have worked on or looked at. Unit testing would be a good idea but I haven't seen this being applied at this level. It seem that everything is currently falling to the test-plan.

Reply:
As a side note, a class with 31 virtual members is concerning.

Reply:
I have fixed this [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/2df0c57cd88f163776d7afc3ea9349f8aba9b18c). However, we do not think that this is a sensible approach as it allows SDK consumers to use unsanitary methods and that these methods can be easily confused with the `Instance` methods that provide good parameter validation. 

To discuss this further, I have created [this issue](https://github.com/project-chip/connectedhomeip/issues/34539). We can discuss and resolve this outside this PR.